File: lintcode/combination-sum.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/combination-sum/

class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum(self, candidates, target):
        # write your code here
        candidates.sort()
        stack = []
        res = []
        for i in xrange(len(candidates)):
            if target-candidates[i] >= 0:
                stack.append([[i], target-candidates[i]])

        while stack:
            path, left = stack.pop()
            if left == 0:
                r = []
                for i in path:
                    r.append(candidates[i])
                res.append(r)
            elif left > 0:
                for i in xrange(path[-1], len(candidates)):
                    if left >= candidates[i]:
                        path_copy = path[:]
                        path_copy.append(i)
                        stack.append([path_copy, left-candidates[i]])

        return res



if __name__ == '__main__':
    s = Solution()
    print s.combinationSum([6,7,2,3], 7)

File: leetcode-2/combination-sum.cc
// Question: https://leetcode.com/problems/combination-sum/

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        helper(candidates, 0, path, ans, target);
        return ans;
    }

    void helper(vector<int>& candidates, int start, vector<int>& path, vector<vector<int>>& ans, int target) {
        if (target == 0) {
            if (!path.empty()) {
                ans.push_back(path);
            }
        }

        for (int i = start; i < candidates.size(); ++i) {
            int n = candidates[i];
            if (target - n >= 0) {
                path.push_back(n);
                helper(candidates, i, path, ans, target - n);
                path.pop_back();
            }
        }
    }
};
